grimoire/core/styling/materials/material.py

```diff
@@ -0,0 +1,295 @@
+from dataclasses import dataclass
+from typing import Callable
+
+from gdpc import Editor, Block
+from glm import ivec3
+
+from grimoire.core.assets.asset import Asset, asset_defaults, default_subtype
+from grimoire.core.noise.hash import recursive_hash, hash_string
+from grimoire.core.noise.seed import Seed
+from grimoire.core.styling.blockform import BlockForm
+from grimoire.core.styling.materials.dithering import (
+    DitheringPattern,
+)
+from grimoire.core.utils.strings import trim_minecraft_label
+
+
+@dataclass
+class MaterialParameters:
+    position: ivec3
+
+    # 0 for darkest, 1 for lightest
+    shade: float
+
+    # 0 for no age, 1 for oldest
+    age: float
+
+    # 0 for no moisture, 1 for most moisture
+    moisture: float
+
+    # dithering pattern, optional
+    dithering_pattern: DitheringPattern | None
+
+
+@dataclass
+class MaterialParameterFunction:
+    shade_func: Callable[[ivec3], float] | None
+    age_func: Callable[[ivec3], float] | None
+    moisture_func: Callable[[ivec3], float] | None
+    dithering_pattern: DitheringPattern | None
+
+    def eval(self, position: ivec3) -> MaterialParameters:
+        return MaterialParameters(
+            position=position,
+            shade=0.5 if self.shade_func is None else self.shade_func(position),
+            age=0 if self.age_func is None else self.age_func(position),
+            moisture=(
+                0 if self.moisture_func is None else self.moisture_func(position)
+            ),
+            dithering_pattern=None,
+        )
+
+
+class Material(Asset):
+    dithering_pattern: DitheringPattern
+
+    def place_block(
+        self,
+        editor: Editor,
+        form: BlockForm,
+        states: dict[str, str] | None,
+        data: str | None,
+        parameters: MaterialParameters,
+    ):
+        pass
+
+    def has_block(self, block: Block) -> bool:
+        pass
+
+    def has_form(self, form: BlockForm) -> bool:
+        pass
+
+    def get_id(self, form: BlockForm, parameters: MaterialParameters) -> str | None:
+        pass
```

Comment:
This is an abstract class where `BasicMaterial` and `CompositeMaterial` provide implementations

Reply:
I can't use python's ABC (_abstract base class_) type here because it relies on Metaclasses to work, which our Asset system does as well- and you can't have two metaclasses at once.

Reply:
That sounds like an unecessary design restriction. Why did you want to use Metaclasses?

Reply:
The Asset system uses Metaclasses so that when every Asset derived class is created, it's registered for serialization. So if we include a `class Book(Asset)` then via the metaclass our serialization system will know to read a json string of `{"type" : "book", ...}` into a `Book` class

Reply:
Might be worth discussing this on a call